### ISSUE TITLE
expand ide:// protocol to do more than just clone

### ISSUE THREAD

**Issue by maxandersen** [starfixdev/starfix#44]
right now we just have `ide://<url to clone>`, but would be interesting to have stuff like `ide://clone?repo=<url>` and `ide://open?path=<url-to-file>` where open would clone if needed and open the file - if repo already cloned just open the file.

Similar could add things like `ide://open?path=<url-to-file>&line=27` to open file and go to line 27.

**Comment by maxandersen** [comment]
related #30  and #7

**Comment by maxandersen** [comment]
idea:

ide://clone?repo=<url>&path=src/main/java/test.java

clone if necessary, open file path


ide://clone?open=https://github.com/starfixdev/starfix/blob/master/cli/installer/windows/installer/starfix.reg

and have "magic parsing" that detects this is a github backed url and do the right clooning and open.
similar can be done for bitbucket, gitlab, etc..

**Comment by maxandersen** [comment]
example patterns for gitlab, github etc. can be found at https://github.com/jbangdev/jbang/blob/478f0c8202ea189362fac80fde5a02f3e7bd1fb5/src/main/java/dev/jbang/util/Util.java

**Comment by maxandersen** [comment]
first step: clone/reuse repo and open file path

second step: handle case where branch does not match or file might not exist locally

**Comment by fahad-israr** [comment]
I tried working with VS Code.
Vs Code provides `code` command to work from Terminal.

```
C:\Users\Fahad Israr>code --help
Visual Studio Code 1.57.1

Usage: code.exe [options][paths...]

To read output from another program, append '-' (e.g. 'echo Hello World | code.exe -')

Options
  -d --diff <file> <file>           Compare two files with each other.
  -a --add <folder>                 Add folder(s) to the last active window.
  -g --goto <file:line[:character]> Open a file at the path on the specified
                                    line and character position.
  -n --new-window                   Force to open a new window.
  -r --reuse-window                 Force to open a file or folder in an
                                    already opened window.
  -w --wait                         Wait for the files to be closed before
                                    returning.
  --locale <locale>                 The locale to use (e.g. en-US or zh-TW).
  --user-data-dir <dir>             Specifies the directory that user data is
                                    kept in. Can be used to open multiple
                                    distinct instances of Code.
  -h --help                         Print usage.
```

- Using the -g argument we can open the file on a specific line and character but the issue is its opening individually.
- We can use --n or -r to decide to open a new window or reuse the existing one but either way the file opens up individually without the rest of the project. 
- If we intend to open the file only then this works fine otherwise will have to think of some other solution.

**Comment by maxandersen** [comment]
code ~/code/myrepo ~/code/myrepo/src/myfile.java

Works as far as I'm aware.

**Comment by fi-javis** [comment]
Oh wow.. that worked !!
 I wasn't aware of that and was just trying things from the vscode cli help .

**Comment by fahad-israr** [comment]
Oh wow.. that worked !!
 I wasn't aware of that and was just trying things from the vscode cli help .

**Comment by fahad-israr** [comment]
- Whenever we write something like this `ide://clone?open=https://github.com/starfixdev/starfix/blob/master/cli/installer/windows/installer/starfix.reg` starfix application will tell us it as an invalid argument. 
- To make this work we can write `starfix clone https://github.com/starfixdev/starfix/blob/master/cli/installer/windows/installer/starfix.reg` .
-    To achieve this we'd have to pre-process this into something like `starfix clone https://git..... `  .Parsing in Java is  easy but I ain't sure how'd I validate it with Picocli . I am currently exploring Picocli docs to see if I can find something helpful there.

**Comment by maxandersen** [comment]
Is this not just because the command does not have first argument defined ? 

i.e. add the following: 

```
@CommandLine.Parameters()
    URI uri;

    @Override
    public void run() {
        System.out.println(uri);
    }
```
and  it will let you pass the URI in as first free argument.

**Comment by fahad-israr** [comment]
We have the first argument defined right ?
Aren't `clone` and `config` the first arguments ?

**Comment by maxandersen** [comment]
Yes but you can other option which will work when it is not clone or config.

**Comment by fahad-israr** [comment]
- This will lead us to have something like this :
```
@Command(name = "clone"){
// Into Clone Command
}

@Command(name = "config) {
// Into Config Command
}
@CommandLine.Parameters()
    URI uri;

    @Override
    public void run() {
        System.out.println(uri);
    }
```
- With this structure picocli will always expect a parameter. 
- For instance, lets  execute : `starfix clone https://git/user/repo` . This is we get:
```
Missing required parameter: '<uri>'
Usage: <main class> [-hV] <uri> [COMMAND]
      <uri>
  -h, --help      Show this help message and exit.
  -V, --version   Print version information and exit.
Commands:
  clone
  config
```

**Comment by fahad-israr** [comment]
I was wondering if Picocli provides something like we have for switch case in Java.
```
switch(var)
case "clone": 
case "config":
default:
{
// ide handler can fit in here
}
```

**Comment by maxandersen** [comment]
Mark the Uri parameter as optional. Then it is not required. See picocli Docs for the syntax.

**Comment by fahad-israr** [comment]
- Yes yes,that's what I was searching for but read through somewhere in the Picocli that single valued parameters are mandatory .
- Recently learnt about setting arity , probably setting it can  make it optional. I am trying with that .

**Comment by fahad-israr** [comment]
I had made changes to the [PR](https://github.com/starfixdev/starfix/pull/59) yesterday.
Have a look if it looks fine.

**Comment by fahad-israr** [comment]
Can close this issue ?
